File: bindings/csharp/examples/MultiAgent.cs
/// Multi-agent coordination: multiple agents collaborating in a shared context.

using Limn.Scp;

static async Task RunAgent(string name, Identity identity, string contextId)
{
    await using var ctx = await Context.JoinAsync(identity, contextId);
    Console.WriteLine($"[{name}] Joined context {contextId}");

    await ctx.SendAsync(System.Text.Encoding.UTF8.GetBytes($"[{name}] reporting in"));

    var count = 0;
    await foreach (var msg in ctx.ReceiveAsync())
    {
        var sender = msg.SenderDid.Length > 16 ? msg.SenderDid[..16] : msg.SenderDid;
        var text = System.Text.Encoding.UTF8.GetString(msg.Content);
        Console.WriteLine($"[{name}] Received from {sender}...: {text}");
        count++;
        if (count >= 2) break;
    }

    Console.WriteLine($"[{name}] Left context");
}

// Create identities for coordinator and two agents
await using var coordinator = await Identity.CreateAsync(custody: "platform");
await using var agentA = await Identity.CreateAsync(custody: "platform");
await using var agentB = await Identity.CreateAsync(custody: "platform");

// Coordinator creates the context
await using var ctx = await Context.CreateAsync(
    coordinator,
    new ContextParams
    {
        Ceiling = ["msg:send", "msg:receive", "tool:invoke"],
        Roles = new Dictionary<string, string[]>
        {
            ["agent"] = ["msg:send", "msg:receive", "tool:invoke"],
        },
        Governance = "single_admin",
    }
);
Console.WriteLine($"Context created: {ctx.ContextId}");

// Mint UCANs for each agent
await Ucan.MintAsync(coordinator, agentA.Did, ["msg:send", "msg:receive"], ctx.ContextId);
await Ucan.MintAsync(coordinator, agentB.Did, ["msg:send", "msg:receive"], ctx.ContextId);

// Run agents concurrently
await Task.WhenAll(
    RunAgent("Agent-A", agentA, ctx.ContextId),
    RunAgent("Agent-B", agentB, ctx.ContextId)
);

File: bindings/java/examples/MultiAgent.java
/** Multi-agent coordination: multiple agents collaborating in a shared context. */

package com.limn.scp.examples;

import com.limn.scp.Context;
import com.limn.scp.Identity;
import com.limn.scp.Types.Message;
import com.limn.scp.Ucan;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Flow;
import java.util.concurrent.CountDownLatch;

public class MultiAgent {

    static void runAgent(String name, Identity identity, String contextId) throws Exception {
        var ctx = Context.join(identity, contextId).join();
        System.out.println("[" + name + "] Joined context " + contextId);

        ctx.send(("[" + name + "] reporting in").getBytes()).join();

        var latch = new CountDownLatch(2);
        ctx.receive().subscribe(new Flow.Subscriber<>() {
            private Flow.Subscription subscription;

            @Override
            public void onSubscribe(Flow.Subscription s) {
                this.subscription = s;
                s.request(Long.MAX_VALUE);
            }

            @Override
            public void onNext(Message msg) {
                var sender = msg.senderDid().length() > 16
                    ? msg.senderDid().substring(0, 16) : msg.senderDid();
                System.out.println("[" + name + "] Received from " + sender + "...: "
                    + new String(msg.content()));
                latch.countDown();
                if (latch.getCount() == 0) subscription.cancel();
            }

            @Override
            public void onError(Throwable t) { t.printStackTrace(); }

            @Override
            public void onComplete() {}
        });
        latch.await();

        ctx.close();
        System.out.println("[" + name + "] Left context");
    }

    public static void main(String[] args) throws Exception {
        // Create identities for coordinator and two agents
        var coordinator = Identity.create("platform").join();
        var agentA = Identity.create("platform").join();
        var agentB = Identity.create("platform").join();

        // Coordinator creates the context
        var ctx = Context.create(coordinator, Map.of(
            "ceiling", List.of("msg:send", "msg:receive", "tool:invoke"),
            "roles", Map.of("agent", List.of("msg:send", "msg:receive", "tool:invoke")),
            "governance", "single_admin"
        )).join();
        System.out.println("Context created: " + ctx.contextId());

        // Mint UCANs for each agent
        Ucan.mint(coordinator, agentA.did(),
            List.of("msg:send", "msg:receive"), ctx.contextId()).join();
        Ucan.mint(coordinator, agentB.did(),
            List.of("msg:send", "msg:receive"), ctx.contextId()).join();

        // Run agents concurrently
        var futureA = CompletableFuture.runAsync(() -> {
            try { runAgent("Agent-A", agentA, ctx.contextId()); }
            catch (Exception e) { throw new RuntimeException(e); }
        });
        var futureB = CompletableFuture.runAsync(() -> {
            try { runAgent("Agent-B", agentB, ctx.contextId()); }
            catch (Exception e) { throw new RuntimeException(e); }
        });
        CompletableFuture.allOf(futureA, futureB).join();

        ctx.close();
        coordinator.close();
        agentA.close();
        agentB.close();
    }
}
